Add strategy selection, recover from failed buy

j-waters opened the pull request:
This kinda supersedes #137 

This adds the multiple active coins strategy and the recovering from failed buys from that PR, as well as the strategy selector

Strategies are loaded dynamically and can be in sub-folders

See the [strategy README](https://github.com/edeng23/binance-trade-bot/blob/64cf06f2afa8c4983f6e63ddf0d68039b4d7e8c3/binance_trade_bot/strategies/README.md) for more information

Pekafu commented on binance_trade_bot/binance_api_manager.py:
```diff
@@ -58,16 +59,24 @@ def retry(self, func, *args, **kwargs):
 
+    @cached(cache=TTLCache(maxsize=2000, ttl=43200))
+    def get_min_notional(self, origin_symbol: str, target_symbol: str):
+        return float(self.get_symbol_filter(origin_symbol, target_symbol, "MIN_NOTIONAL")["minNotional"])
```
Should this be qty in bridge currency? Since some coins are more than 50kUSD per coin and some are less than cents.

j-waters replied:
This does return qty in bridge currency. It's used [here](https://github.com/edeng23/binance-trade-bot/pull/204/files#diff-aa0c7b55165bbdbd7f5282b0a3b3b86719ca12ae39dab02fdb7ce01017421be1R20)

Pekafu commented on README.md:
````diff
@@ -112,6 +114,9 @@ pip install pre-commit
 pre-commit install
 ```
 
+The scouting algorithm is unlikely tobe changed. If you'd like to contribute an alternative
````
```suggestion
The scouting algorithm is unlikely to be changed. If you'd like to contribute an alternative
```

Pekafu commented on binance_trade_bot/auto_trader.py:
```diff
@@ -20,35 +18,36 @@ def __init__(self, binance_manager: BinanceAPIManager, database: Database, logge
         self.logger = logger
         self.config = config
 
+    def initialise(self):
```
```suggestion
    def initialize(self):
```

Pekafu reviewed (commented):
Seems very nice way to branch into multiple strategies. I'll try to test this in upcoming days.

I'm wondering if introducing custom strategies should be left for later if we get websockets at some time to get rid of the scouting in general. Or atleast put disclamer in the strategies readme that things might get quite an overhaul in upcoming weeks.

Pekafu commented on binance_trade_bot/crypto_trading.py:
```diff
@@ -16,16 +16,19 @@ def main():
 
-    trader.initialize_trade_thresholds()
-    trader.initialize_current_coin()
+    trader.initialise()
```
```suggestion
    trader.initialize()
```

j-waters commented:
I always though websockets would only be used to replace waiting for trades to complete, not scouting, but I could be wrong! I don't think there's too much harm merging this now, I can put a disclaimer if people think it's necessary.

j-waters commented:
Also thinking about it some more, do we actually want 'official' alternative strategies? I added the one I've been using because I do think it's an improvement but it might just add more development overhead, and I know @edeng23 is busy.

Some options:

1. If people have custom strategies that are well polished and useful, we make them 'official' and add them to the repo, like I've done with mine
2. Similar to the above, but instead we use git submodules, so that the developer of the strategy can work on it without having to constantly send pull requests. The submodule would still have to be updated however (could possibly be done automatically)
3. We add a strategy registry - a JSON file here that contains a list of strategy names and their corresponding git repos. When configuring which strategy to use, the bot will automatically pull the selected strategy. Adding strategies to the registry will still require a pull request but once they've been added the original developer is free to update it without having to involve the team here

Now I really like 3, and would be happy to build a system like that, but it does come with issues - there's the possibility of the 3rd party strategy turning malicious

2 is also good, but still requires someone with write access to this repo to update the submodule if there are any changes. We could maybe add something that automatically does that, but then we run into the same problem as with 3

1 would be fine although I probably wouldn't bother sending a PR every time I changed something.

j-waters commented:
Also something to think about is configuration for strategies - probably extra sections in the config file and then a custom strategy config class. I'll try and think about how to make the API as clean as possible

Pekafu commented:
> 3. We add a strategy registry - a JSON file here that contains a list of strategy names and their corresponding git repos. When configuring which strategy to use, the bot will automatically pull the selected strategy. Adding strategies to the registry will still require a pull request but once they've been added the original developer is free to update it without having to involve the team here

I also like the 3rd option, perhaps even having own public (multiple collabs) repo for the strategies

edeng23 commented:
looks amazing! I really love the liberty that the modularization introduced. I'll start running tests later today, and hopefully merge very soon :)

cornytrace commented:
I'm getting an error while trying this pr with a fresh database:

`2021-03-19 09:32:40,766 - crypto_trading_logger - INFO - Initializing <DOT> vs <UNI>
2021-03-19 09:32:40,898 - crypto_trading_logger - INFO - Setting initial coin to DASH
2021-03-19 09:32:41,041 - crypto_trading_logger - INFO - Purchasing <DASH> to begin trading
2021-03-19 09:32:44,137 - crypto_trading_logger - INFO - BUY QTY 1.80141
2021-03-19 09:32:44,369 - crypto_trading_logger - INFO - {'symbol': 'DASHUSDT', 'orderId': 575488572, 'orderListId': -1, 'clientOrderId': 'hMBMW1RfJiCQULFlIfSBH5', 'transactTime': 1616142764257, 'price': '228.85000000', 'origQty': '1.80141000', 'executedQty': '1.80141000', 'cummulativeQuoteQty': '411.87438240', 'status': 'FILLED', 'timeInForce': 'GTC', 'type': 'LIMIT', 'side': 'BUY', 'fills': [{'price': '228.64000000', 'qty': '1.80141000', 'commission': '0.00180141', 'commissionAsset': 'DASH', 'tradeId': 22788643}]}
2021-03-19 09:32:44,803 - crypto_trading_logger - INFO - {'symbol': 'DASHUSDT', 'orderId': 575488572, 'orderListId': -1, 'clientOrderId': 'hMBMW1RfJiCQULFlIfSBH5', 'price': '228.85000000', 'origQty': '1.80141000', 'executedQty': '1.80141000', 'cummulativeQuoteQty': '411.87438240', 'status': 'FILLED', 'timeInForce': 'GTC', 'type': 'LIMIT', 'side': 'BUY', 'stopPrice': '0.00000000', 'icebergQty': '0.00000000', 'time': 1616142764257, 'updateTime': 1616142764257, 'isWorking': True, 'origQuoteOrderQty': '0.00000000'}
2021-03-19 09:32:44,804 - crypto_trading_logger - INFO - Bought DASH
2021-03-19 09:32:44,925 - crypto_trading_logger - INFO - Ready to start trading
2021-03-19 09:33:06,023 - crypto_trading_logger - ERROR - Error while scouting...
Traceback (most recent call last):
  File "/mnt/data/binance-trade-bot/binance_trade_bot/scheduler.py", line 25, in _run_job
    super()._run_job(job)
  File "/home/cornee/.local/lib/python3.8/site-packages/schedule/__init__.py", line 147, in _run_job
    ret = job.run()
  File "/home/cornee/.local/lib/python3.8/site-packages/schedule/__init__.py", line 491, in run
    ret = self.job_func()
  File "/mnt/data/binance-trade-bot/binance_trade_bot/strategies/default_strategy.py", line 41, in scout
    self._jump_to_best_coin(current_coin, current_coin_price, all_tickers)
  File "/mnt/data/binance-trade-bot/binance_trade_bot/auto_trader.py", line 129, in _jump_to_best_coin
    ratio_dict = self._get_ratios(coin, coin_price, all_tickers)
  File "/mnt/data/binance-trade-bot/binance_trade_bot/auto_trader.py", line 116, in _get_ratios
    transaction_fee = self.manager.get_fee(pair.from_coin, self.config.BRIDGE, True) + self.manager.get_fee(
  File "/mnt/data/binance-trade-bot/binance_trade_bot/binance_api_manager.py", line 45, in get_fee
    fee_amount_bnb = fee_amount * self.get_market_ticker_price(origin_coin + Coin("BNB"))
TypeError: unsupported operand type(s) for *: 'float' and 'NoneType'`

This error repeats every ~15 seconds.

cornytrace commented:
I discovered my previous error is because this branch is not current with master.
Rebasing this branch on master resolves my error.

cornytrace commented on binance_trade_bot/strategies/multiple_coins.py:
```diff
@@ -0,0 +1,31 @@
+
+            # Display on the console, the current coin+Bridge, so users can see *some* activity and not think the bot
+            # has stopped. Not logging though to reduce log size.
+            self.logger.info(f"Scouting for best trades. Current ticker: {coin + self.config.BRIDGE} ")
```
You are using logging where the comments indicate you should not.
Logging here instead of using print will result in a lot of logspam.

cornytrace commented on supported_coin_list:
```diff
@@ -10,7 +10,6 @@ ICX
 IOTA
 NEO
 OMG
-ONT
```
Is this intentional? If not, please take a look at my pr #220 which is made to address such problems  ;-)

cornytrace commented on binance_trade_bot/strategies/__init__.py:
```diff
@@ -1,5 +1,6 @@
 
 def get_strategy(name):
-    return next((S for S in [DefaultStrategy] if S.__name__ == name), None)
+    return next((S for S in [DefaultStrategy, MultipleActiveCoinsStrategy] if S.__name__ == name), None)
```
This implies that new strategies would need to be manually added to this file.
It would be better if we use importlib.import_module() to import the strategy file automatically.

cornytrace reviewed (changes requested):
While this PR is working fine for me after the rebase, I have found some minor issues that should be addressed.

j-waters commented:
Rebased

edeng23 commented:
Thanks, testing again.

pwnfoo commented:
^^testing out this one and watching this thread :)

j-waters commented:
There is an argument to be made of what should be in the `AutoTrader` class and what should be in the `DefaultStrategy` class (I did move quite a lot over). I'd be happy to move stuff back

cornytrace commented:
> There is an argument to be made of what should be in the `AutoTrader` class and what should be in the `DefaultStrategy` class (I did move quite a lot over). I'd be happy to move stuff back

Ideally only the core business logic should be in DefaultStrategy, any utility function that would be even remotely useful in other strategies should be in AutoTrader to prevent code duplication. I see nothing that should be moved out of DefaultStrategy.

One thing that this does bring to my mind while reading DefaultStrategy is that not all strategies will have a concept of a current_coin. It would make sense to namespace all strategy-specific configuration and db entries, and to define these user.cfg entries and db structures in the strategy file itself. I do however believe that this can be done in a new PR, as merging this PR quickly means other PRs can start rebasing on these changes. (I'm particularly looking forward to #211)

j-waters commented:
Good points. `bridge_scout` is also called in `DefaultStrategy` which is why I left it in `AutoTrader`, to reduce duplication.

Namespacing configuration I definitely agree with, I experimented a bit with how we'd do it but never finalised anything. With `ConfigParser`, we could have a different heading for each strategy, named after the strategy. Then when you're defining your strategy, you also define a config class that extends the original config class and pass that in somewhere. Not too sure on the specifics, I might leave it to another PR where we could also improve the `Config` class, which has a lot of repetition.

Namespacing the database is a very interesting point. It probably will be necessary eventually, but I'm not sure how we'd begin! The issue with an SQL database is that changes to the schema are a bit of a pain. I can't imagine someone developing their own strategy also going to the trouble of providing proper database migration scripts. So maybe we'll have to come up with another solution if people want to add persistence to their strategies on top of what is already provided. I kind of imagined that most strategies would be based off the original algorithm, and so people would want to keep their `current_coin` and ratios as they switch between them. Although if one strategy changed the way they calculate ratios, that would break all the others! Tricky!

But yeah I think those are issues for another PR, and we should make it abundantly clear that custom strategies are in an early alpha state

cornytrace commented:
For first party / in-repo strategies we can choose to only accept PRs where migration is handled gracefully, or at least it deletes and recreates its tables if it will not affect trade performance. If it does affect trade performance, I assume the developer will have financial incentive to at least migrate their own data, at which point most of the work for doing it automatically is already done. 
For third party / external strategies: Who cares not our problem lmao. If all strategy tables are prefixed with strat_<strategy name here>_ we could add a command to clear the db for the specified strategy if needed. 
I do not see where bridge_scout is used in default_strategy, could you point it out to me?

edeng23 commented:
Previous tests passed successfully. Testing again due to latest commit.

pwnfoo commented:
Been testing this out for a while and works great for me! Nice one @j-waters ;)

cornytrace commented:
Could someone give an example where bridge_scout would actually do something, at least in the DefaultStrategy? Wouldn't all the bridge coin be used up jumping from one coin to the other? And the minuscule amount of bridge coin that remains not worth the transaction fee to buy another coin with?

j-waters commented:
@cornytrace It's there to recover from a failed or cancelled buy.

Scenario:
1. Bot decides to jump from ETH to XLM
2. Bot sells ETH so we now have a bunch of USDT
3. We attempt to buy, but something goes wrong
3a. Say the bot is accidentally closed between selling and buying. Now we're stuck on the bridge currency and the bot can't recover until the user manually rebuys ETH, or buys another coin and changes the current coin in the database. In both cases, the coin the user manually buys might not be the optimal one to jump to
3b. Say the bot fails to buy for some other reason but keeps running. [It will just keep attempting to buy that same coin forever](https://github.com/edeng23/binance-trade-bot/blob/master/binance_trade_bot/auto_trader.py#L33), even if doing so is now a bad decision

I should really have put the `bridge_scout` method in another pull request, but this PR is just a rewrite of https://github.com/edeng23/binance-trade-bot/pull/137 so I didn't really think about it. I talked to @edeng23 about this earlier and if he wants me to remove it or just have it in the multiple coins strategy (where it has more use) I'd be fine with that

cornytrace commented:
I understand now, thanks for explaining. I have experienced scenario 3b for myself, so I'm glad this does something about it. It seems totally appropriate to use in both strategies.
